File: sdks/ruby/lib/walinko/configuration.rb
# frozen_string_literal: true

module Walinko
  # Immutable configuration captured by `Walinko::Client.new`. Validates
  # required fields and applies defaults.
  class Configuration
    DEFAULT_BASE_URL     = 'https://api.walinko.com'
    DEFAULT_TIMEOUT      = 30
    DEFAULT_OPEN_TIMEOUT = 10
    DEFAULT_MAX_RETRIES  = 2

    attr_reader :api_key, :base_url, :timeout, :open_timeout,
                :max_retries, :logger

    # @param api_key      [String]   required, e.g. "walk_live_<keyId>.<secret>"
    # @param base_url     [String]   defaults to "https://api.walinko.com"
    # @param timeout      [Integer]  per-request read timeout in seconds
    # @param open_timeout [Integer]  TCP/TLS connection-setup timeout in seconds
    # @param max_retries  [Integer]  retries on idempotent failures (network, 429, 5xx)
    # @param logger       [#info,#warn,#error] optional structured logger
    def initialize(api_key:,
                   base_url:     DEFAULT_BASE_URL,
                   timeout:      DEFAULT_TIMEOUT,
                   open_timeout: DEFAULT_OPEN_TIMEOUT,
                   max_retries:  DEFAULT_MAX_RETRIES,
                   logger:       nil)
      raise ArgumentError, 'api_key is required' if api_key.nil? || api_key.to_s.empty?
      raise ArgumentError, 'base_url is required' if base_url.nil? || base_url.to_s.empty?
      raise ArgumentError, 'timeout must be > 0' if timeout.to_i <= 0
      raise ArgumentError, 'open_timeout must be > 0' if open_timeout.to_i <= 0
      raise ArgumentError, 'max_retries must be >= 0' if max_retries.to_i.negative?

      @api_key      = api_key.to_s
      @base_url     = base_url.to_s.sub(%r{/+$}, '')
      @timeout      = timeout.to_i
      @open_timeout = open_timeout.to_i
      @max_retries  = max_retries.to_i
      @logger       = logger
    end
  end
end

File: sdks/php/src/Configuration.php
<?php

declare(strict_types=1);

namespace Walinko;

use Psr\Http\Client\ClientInterface as HttpClientInterface;
use Psr\Http\Message\RequestFactoryInterface;
use Psr\Http\Message\StreamFactoryInterface;
use Psr\Log\LoggerInterface;

/**
 * Immutable configuration captured by `Walinko\Client::__construct()`.
 *
 * Validates required fields and applies defaults. PSR-18 / PSR-17
 * dependencies are auto-discovered via `php-http/discovery` when the
 * caller doesn't inject them explicitly.
 *
 * Note on `timeout`: PSR-18 has no standard transport-level timeout
 * setting. The SDK uses `timeout` to seed `Messages::waitUntilDone`'s
 * default deadline; if you want a per-request timeout on the wire,
 * inject a pre-configured PSR-18 client (e.g. Guzzle / Symfony with
 * `timeout` / `read_timeout` set in the constructor).
 */
final class Configuration
{
    public const DEFAULT_BASE_URL = 'https://api.walinko.com';
    public const DEFAULT_TIMEOUT = 30;
    public const DEFAULT_MAX_RETRIES = 2;

    public readonly string $apiKey;
    public readonly string $baseUrl;
    public readonly int $timeout;
    public readonly int $maxRetries;
    public readonly ?HttpClientInterface $httpClient;
    public readonly ?RequestFactoryInterface $requestFactory;
    public readonly ?StreamFactoryInterface $streamFactory;
    public readonly ?LoggerInterface $logger;

    /**
     * @param array{
     *     api_key: string,
     *     base_url?: string,
     *     timeout?: int,
     *     max_retries?: int,
     *     http_client?: HttpClientInterface,
     *     request_factory?: RequestFactoryInterface,
     *     stream_factory?: StreamFactoryInterface,
     *     logger?: LoggerInterface
     * } $options
     */
    public function __construct(array $options)
    {
        $apiKey = $options['api_key'] ?? '';
        if (!\is_string($apiKey) || $apiKey === '') {
            throw new \InvalidArgumentException('api_key is required');
        }

        $baseUrl = $options['base_url'] ?? self::DEFAULT_BASE_URL;
        if (!\is_string($baseUrl) || $baseUrl === '') {
            throw new \InvalidArgumentException('base_url must be a non-empty string');
        }

        $timeout = (int) ($options['timeout'] ?? self::DEFAULT_TIMEOUT);
        if ($timeout <= 0) {
            throw new \InvalidArgumentException('timeout must be > 0');
        }

        $maxRetries = (int) ($options['max_retries'] ?? self::DEFAULT_MAX_RETRIES);
        if ($maxRetries < 0) {
            throw new \InvalidArgumentException('max_retries must be >= 0');
        }

        $this->apiKey = $apiKey;
        $this->baseUrl = rtrim($baseUrl, '/');
        $this->timeout = $timeout;
        $this->maxRetries = $maxRetries;
        $this->httpClient = $options['http_client'] ?? null;
        $this->requestFactory = $options['request_factory'] ?? null;
        $this->streamFactory = $options['stream_factory'] ?? null;
        $this->logger = $options['logger'] ?? null;
    }
}
